### PR TITLE
Alternate to #77 - Fix OpenSSL related crashes on OSX and Arch

### PR DESCRIPTION
**Pull request by posita**
Avoids segfaults on OS X. Expands approach from #77 of manually specifying argument and return types for all `ctype` calls (except one; see updated [`key.py#L170`](/petertodd/python-bitcoinlib/pull/79/files#diff-dd0294243fd03d28503c0f83d7715492R170) and accompanying note).

As an aside, see #80 for adding tox support to automate testing on multiple versions and to see whether certain `ctypes` calls are being hit by the various tests.

### CODE REVIEW

**Comment by posita** [comment]
As an aside, see #80 for adding tox support to automate testing on multiple versions and to see whether certain `ctypes` calls are being hit by the various tests.

**Comment by bitstein** [comment]
Ack.

Confirmed for passing tests on OSX 10.9.5:
```
(venv) python-bitcoinlib2 git:(d1584ff) $ python -m unittest discover && python3 -m unittest discover
......................................................................................................
----------------------------------------------------------------------
Ran 102 tests in 1.999s

OK
......................................................................................................
----------------------------------------------------------------------
Ran 102 tests in 2.027s

OK
```

File: bitcoin/core/key.py [hunk 1]
```diff
@@ -35,15 +35,142 @@
...
+_ssl.i2d_ECPrivateKey.restype = ctypes.c_int
+_ssl.i2d_ECPrivateKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.i2o_ECPublicKey.restype = ctypes.c_void_p
+_ssl.i2o_ECPublicKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p]
+
+# Why do tests fail when uncommenting this? What other errors are being
+# masked (if any) by weird ctypes usages or misuse of ctypes.c_void_p?
+# _ssl.o2i_ECPublicKey.restype = ctypes.c_void_p
+# _ssl.o2i_ECPublicKey.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_long]
```
Comment (posita) [inline bitcoin/core/key.py]: The thing that makes me nervous of this approach in general is that it uses `c_void_p` almost everywhere where a pointer is used in the underlying API (including places which expect pointers to pointers). If I uncomment the preceding two lines, I get the following errors when running unit tests, which I can't explain:

```sh
% python -m unittest discover
...................F..............................................................................F...
======================================================================
FAIL: test (bitcoin.tests.test_key.Test_CPubKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../python-bitcoinlib/bitcoin/tests/test_key.py", line 27, in test
    T('', False, False, False)
  File ".../python-bitcoinlib/bitcoin/tests/test_key.py", line 24, in T
    self.assertEqual(key.is_fullyvalid, is_fullyvalid)
AssertionError: True != False

======================================================================
FAIL: test_from_invalid_pubkeys (bitcoin.tests.test_wallet.Test_P2PKHBitcoinAddress)
Create P2PKHBitcoinAddress's from invalid pubkeys
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../python-bitcoinlib/bitcoin/tests/test_wallet.py", line 196, in test_from_invalid_pubkeys
    P2PKHBitcoinAddress.from_pubkey(x(''))
AssertionError: CBitcoinAddressError not raised

----------------------------------------------------------------------
Ran 102 tests in 1.670s

FAILED (failures=2)
```

Comment (posita) [inline bitcoin/core/key.py]: Investigating further, it appears that `o2i_ECPublicKey` [calls `EC_POINT_oct2point`](https://github.com/openssl/openssl/blob/74924dcb3802640d7e2ae2e80ca6515d0a53de7a/crypto/ec/ec_asn1.c#L1250), which in turn either [calls `ec_GFp_simple_oct2point`](https://github.com/openssl/openssl/blob/6218a1f57e7e25a6b9a798f00cf5f0e56a02ff31/crypto/ec/ec_oct.c#L180) (which [should reject zero-length keys](https://github.com/openssl/openssl/blob/1ae3fdbe6a7e78eef4d7678cb08ff468481c24ad/crypto/ec/ecp_oct.c#L334)), or it [calls something else](https://github.com/openssl/openssl/blob/6218a1f57e7e25a6b9a798f00cf5f0e56a02ff31/crypto/ec/ec_oct.c#L191) that I haven't yet tracked down.

Comment (petertodd) [inline bitcoin/core/key.py]: Ok, that's promising.

@posita Can you check if https://github.com/petertodd/python-bitcoinlib/pull/78 makes the above issue go away? (IE uncommenting those two lines?)

Comment (posita) [inline bitcoin/core/key.py]: @petertodd, unfortunately, it doesn't. [Merging #78 into this fix and uncommenting the above two lines](https://github.com/posita/python-bitcoinlib/commits/posita/merge-posita-fix_osx_and_arch_crashes-simcity4242-patch_1) results in the same two test failures. :disappointed:

The (sort of?) good news is that even without the ctypes specification, calls to `o2i_ECPublicKey` don't segfault. Whether they're doing what they're supposed to is another matter which I *hope* is adequately covered in the unit tests. However, the fact that providing a ctypes spec for `o2i_ECPublicKey` causes unit test failures still makes me nervous about using a similar technique with the other OpenSSL call-out functions. Unless the unit tests are wrong, but that doesn't appear to be the case. I don't think zero-length octal-encoded public keys should be considered valid.

I guess the alternative would be to write (or port) unit tests for each OpenSSL function to make sure the ctypes specifications didn't screw things up. That's probably well beyond the scope of what was originally intended here.

Comment (jashug) [inline bitcoin/core/key.py]: I found the bug. In `CPubKey.__new__`, we have the line `self.is_fullyvalid = _cec_key.set_pubkey(self) != 0`. Without these declarations, pointers are returned as integers, and can be compared to 0. With these declarations, pointers are almost always returned as integers, except when they are `NULL` (which used to be 0). Then Python translates them to `None`, and `None != 0`. Therefore `is_fullyvalid` gets set to `True`. This can be easily fixed by using `self.is_fullyvalid = bool(_cec_key.set_pubkey(self))`, which works whether 0 or `None` is returned. Tested on my computer, this fixes both test failures.

Comment (posita) [inline bitcoin/core/key.py]: Hah! Brilliant! :bowtie: Working on a PR update now....

P.S. @jashug, do you want a job? :wink:

File: bitcoin/core/key.py [hunk 2]
```diff
@@ -35,26 +35,151 @@
...
+_ssl.EC_GROUP_get_degree.restype = ctypes.c_int
+_ssl.EC_GROUP_get_degree.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_GROUP_get_order.restype = ctypes.c_int
+_ssl.EC_GROUP_get_order.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_void_p]
+
+_ssl.EC_KEY_free.restype = None
+_ssl.EC_KEY_free.argtypes = [ctypes.c_void_p]
+
+_ssl.EC_KEY_new_by_curve_name.restype = ctypes.c_void_p
```
Comment (posita) [inline bitcoin/core/key.py]: We set the return type here, but ...

File: bitcoin/core/key.py [hunk 3]
```diff
@@ -67,7 +192,10 @@ class CECKey:
     POINT_CONVERSION_UNCOMPRESSED = 4
 
     def __init__(self):
-        self.k = _ssl.EC_KEY_new_by_curve_name(_NID_secp256k1)
+        # WTF: not sure why the cast is necessary here, but we get back an
+        # int without it, even though we set
+        # _ssl.EC_KEY_new_by_curve_name.restype above
+        self.k = ctypes.c_void_p(_ssl.EC_KEY_new_by_curve_name(_NID_secp256k1))
```
Comment (posita) [inline bitcoin/core/key.py]: ... if we don't have this "cast", `EC_KEY_new_by_curve_name` seems to return an `int`<sup>1</sup>, which means the various invocations of `ctypes.byref(self.k)` will throw an exception. The "cast" fixes this, but (at least in theory) it shouldn't be necessary.

--

<sup>1</sup> Although it appears that on 64-bit machines the `int` does have a high word, so that's at least something. :grimacing:

**Comment by posita** [comment]
Added @jashug's fix and rebased to master.

```sh
% python -m unittest discover && python3.4 -m unittest discover
......................................................................................................
----------------------------------------------------------------------
Ran 102 tests in 2.857s

OK
......................................................................................................
----------------------------------------------------------------------
Ran 102 tests in 3.359s

OK
```

File: bitcoin/core/key.py (hunk 3 above)
Comment (jashug) [inline bitcoin/core/key.py]: You are slightly misunderstanding the ctypes system (though note that I am no master). The return type of C functions is always a Python type. Even when you set `restype = c_void_p`, this only means that you get back a (64 bit) `int` or `None`. However, if you set both argument types and return types this way, you *shouldn't* run into conversion issues.

The docs are https://docs.python.org/2/library/ctypes.html#fundamental-data-types, though I don't feel they spell it out clear enough. What comes out is the 'Python type' they mention.

Comment (posita) [inline bitcoin/core/key.py]: Ah, okay, that makes perfect sense then. From <https://docs.python.org/2/library/ctypes.html#ctypes.byref>:

> Returns a light-weight pointer to obj, which must be an instance of a ctypes type.

Meaning that in Python, `ctypes.byref(thing)` is (roughly) equivalent to `&thing` in C. It's used here in places like:

```py
... _ssl.d2i_ECPrivateKey(ctypes.byref(self.k), ...)
```

Because `self.k` represents a `EC_KEY*` and the C [signature for d2i_ECPrivateKey](https://www.openssl.org/docs/manmaster/crypto/d2i_ECPrivate_key.html) is:

```c
... d2i_ECPrivateKey(EC_KEY **key, ...)
```

Hopefully that all makes sense.

In other words, now that you pointed out (and I bothered to read) the `ctypes` docs, the "cast" looks entirely appropriate. It was my mistake to think that it wasn't. I'll clean up the code (and my comment).

**Comment by petertodd** [comment]
Looks good to me guys, thanks!
